sql: migrate prepared statements during session migration

rafiss opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/76398

See individual commits. The last one contains the business logic changes,
and the rest are just refactors.

Release note (sql change): The crdb_internal.serialize_session and
crdb_internal.deserialize_session now can handle prepared statements.
When deserializing, any prepared statements that existed when the
session was serialized are re-prepared. It is an error to re-prepare a
statement if the current session already has a statement with that name.

rafiss committed:
sql: migrate prepared statements during session migration

Release note (sql change): The crdb_internal.serialize_session and
crdb_internal.deserialize_session now can handle prepared statements.
When deserializing, any prepared statements that existed when the
session was serialized are re-prepared. It is an error to re-prepare a
statement if the current session already has a statement with that name.

rafiss committed:
sql: move {de}serialze_session to EvalPlanner

This will allow us to more easily call into the code to prepare
statements when deserializing.

Release note: None

rafiss committed:
sql: expose an interface for preparing statements using evalCtx

Release note: None

rafiss committed:
sql: remove CheckCanBecomeUser from EvalPlanner

After the recent refactor for deserialize_session, it no longer needs to
be exposed.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76399)
<!-- Reviewable:end -->

otan reviewed (approved):
changes lgtm.

one thing i am curious about  - what happens to prepared statements if a schema change happened and the types change? in turn, does `AddPreparedStatement` during a session migration potentially change the result of the prepared statement producing an entirely different statement? is it worth testing?

andy-kimball commented:
CC @RaduBerinde to make sure to take a look at this.

rafiss committed:
sql: restore invalid prepared statements in session migration

A schema change can make an existing prepared statement invalid. Rather
than returning an error when deserializing that invalid statement, now
the error will be returned only when trying to use the statement. This
makes a session migration even more transparent to the end user.

Release note: None

rafiss commented:
> what happens to prepared statements if a schema change happened and the types change? 

This is the behavior without any session migration involved:
```
exec
CREATE TABLE t(a INT4, b TEXT, c DATE)
----

wire_prepare s2
INSERT INTO t VALUES($1, $2, $3)
----

exec
ALTER TABLE t DROP COLUMN c
----

wire_exec s2 1 cat 2022-02-10
----
ERROR: INSERT has more expressions than target columns, 3 expressions for 2 targets (SQLSTATE 42601)
```

> in turn, does AddPreparedStatement during a session migration potentially change the result of the prepared statement producing an entirely different statement? is it worth testing?

yeah we definitely should test. given the above, i guess the two options are to not serialize a prepared statement if it's not valid at the time of serialization, or to ignore certain types of prepare errors at the time of deserialization.

rafiss commented:
We could also keep the behavior as-is and just error out. (The user-impact is that their client would get a hard disconnect rather than a graceful session migration.)

otan commented:
happy with erroring out. also happy for you to add this test in a separate PR.

rafiss commented:
confirmed the behavior for prepared statements after a schema change at the last multi-tenant team meeting with signoff from @JeffSwenson so going to merge this.

bors r=otan

rafiss commented:
bors r=otan

jaylim-crl reviewed (commented):
I believe this will fail: https://github.com/cockroachdb/cockroach/blob/696cdd5daa339a700f77ddbfd4759dd054c60b2d/pkg/ccl/testccl/sqlccl/show_transfer_state_test.go#L199-L225

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76399#-:-MwZcLo59HhPFmbTPnpq:b6mo6uk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jaylim-crl and @rafiss)

___
*[pkg/sql/types/types.go, line 1746 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76399#-MwZci9l0KawYQlhuT3w:-MwZci9l0KawYQlhuT3x:bw8g3xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad02cf51934b17dfcb960018c3532083896ca2f5/pkg/sql/types/types.go#L1746)):*
> // messages
//and also to produce the output of SHOW CREATE.


nit: typo here?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
ah, whoops. fixing

bors r-

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76399#-:-MwZo6dICEYw-LNI5mCG:b5imta2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jaylim-crl)

___
*[pkg/sql/types/types.go, line 1746 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76399#-MwZci9l0KawYQlhuT3w:-MwZo4kI0WQvQuhgyUts:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad02cf51934b17dfcb960018c3532083896ca2f5/pkg/sql/types/types.go#L1746)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

> // messages
//and also to produce the output of SHOW CREATE.


nit: typo here?
</blockquote></details>

fixed


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
@jaylim-crl i edited that test, in case you want to take a look

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76399#-:-MwZrGN9ELTS5lZy7EOz:bds3u92)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jaylim-crl)



<!-- Sent from Reviewable.io -->

otan commented:
do you think this is causing the `TestShowTransferState/errors/prepared_statements` flake?

jaylim-crl reviewed (commented):
> @jaylim-crl i edited that test, in case you want to take a look

Thanks. The test LGTM.

> Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 4 of 4 files at r3, 13 of 16 files at r4, 4 of 4 files at r6, 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76399#-:-MwbUFMw8uGBVdlBwZjl:b-w0m6km)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

rafiss commented:
tftrs!

bors r=otan

rafiss commented:
the failures all seem to be flakes as far as i can tell..

bors r=otan

rafiss commented:
yet another flake..

bors r=otan

otan commented:
congrats on winning the CI lottery!
